File: 1570-final-prices-with-a-special-discount-in-a-shop/solution.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        vector<int> s;
        for (int i = 0; i < prices.size(); ++i) {
            while (s.size() && prices[s.back()] >= prices[i]) {
                prices[s.back()] -= prices[i];
                s.pop_back();
            }
            s.push_back(i);
        }

        return prices;
    }
};

File: 1570-final-prices-with-a-special-discount-in-a-shop/solution.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        for i in range(len(prices)):
            # Look for first smaller or equal price that comes after current item
            for j in range(i + 1, len(prices)):
                if prices[j] <= prices[i]:
                    # Apply discount by subtracting prices[j] from current price
                    prices[i] = prices[i] - prices[j]
                    break

        return prices
